File: Binary Tree Maximum Path Sum.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void help(TreeNode *root, int &curmax, int &maxSum) {
        if (root == NULL) {
            curmax = 0;
            return;
        }
        
        int lmax, rmax;
        help(root->left, lmax, maxSum);
        help(root->right, rmax, maxSum);
        curmax = max(max(lmax, rmax), 0) + root->val;
        maxSum = max(maxSum, max(curmax, lmax+rmax+root->val));
    }
    int maxPathSum(TreeNode *root) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        int maxSum = INT_MIN, curmax;
        help(root, curmax, maxSum);
        return maxSum;
    }
};

File: Binary Tree Maximum Path Sum.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int MaxPathSum(TreeNode root) {
        var curMax = 0;
        return MaxPathSum(root, out curMax);
    }
    
    private int MaxPathSum(TreeNode root, out int curMax) {
        if (root == null) {
            curMax = 0;
            return int.MinValue;
        }
        
        var leftMax = 0;
        var rightMax = 0;
        var maxSum = Math.Max(MaxPathSum(root.left, out leftMax), MaxPathSum(root.right, out rightMax));
        curMax = Math.Max(Math.Max(leftMax, rightMax), 0) + root.val;
        
        return Math.Max(maxSum, Math.Max(curMax, leftMax + rightMax + root.val));
    }
}
